File: backend/app/services/habit_completion_reporter.py
"""
Habit Completion Reporter

Service for processing habit completions from Slack interactions.

Requirements:
- 2.1: WHEN Supabaseへのリクエストが接続エラーで失敗する THEN Retry_Logic SHALL 指数バックオフで最大3回リトライする
- 2.3: WHEN 全てのリトライが失敗する THEN Retry_Logic SHALL 最終エラーを適切にログ出力して例外を発生させる
"""

import logging
from typing import Optional, List, Dict, Any, Tuple
from datetime import date, datetime, timedelta
from difflib import SequenceMatcher
from supabase import Client

from app.utils.retry import with_retry

logger = logging.getLogger(__name__)


class HabitCompletionReporter:
    """Service for handling habit completions via Slack."""

    def __init__(self, supabase: Client):
        self.supabase = supabase

    async def complete_habit_by_name(
        self,
        owner_id: str,
        habit_name: str,
        source: str = "slack",
        owner_type: str = "user",
    ) -> Tuple[bool, str, Optional[Dict]]:
        """
        Find and complete a habit by name.
        
        Args:
            owner_id: User ID
            habit_name: Name of the habit to complete
            source: Source of completion (default: "slack")
            owner_type: Type of owner (default: "user")
            
        Returns:
            Tuple of (success, message, habit_data)
        """
        # Find habit by name (case-insensitive)
        habits = await self._get_habits_by_owner(owner_type, owner_id)
        
        # Try exact match first
        habit = next(
            (h for h in habits if h["name"].lower() == habit_name.lower()),
            None
        )
        
        # Try partial match if no exact match
        if not habit:
            habit = next(
                (h for h in habits if habit_name.lower() in h["name"].lower()),
                None
            )
        
        if not habit:
            # Find similar habits for suggestions
            similar = self._find_similar_habits(habit_name, habits)
            if similar:
                return False, f"Habit '{habit_name}' not found", {"suggestions": similar}
            return False, f"Habit '{habit_name}' not found", None

        return await self.complete_habit_by_id(
            owner_id, habit["id"], source, owner_type
        )

    async def complete_habit_by_id(
        self,
        owner_id: str,
        habit_id: str,
        source: str = "slack",
        owner_type: str = "user",
    ) -> Tuple[bool, str, Optional[Dict]]:
        """
        Complete a habit by ID.
        
        Args:
            owner_id: User ID
            habit_id: ID of the habit to complete
            source: Source of completion (default: "slack")
            owner_type: Type of owner (default: "user")
            
        Returns:
            Tuple of (success, message, habit_data)
        """
        # Get habit details
        habit = await self._get_habit_by_id(habit_id)
        if not habit:
            return False, "Habit not found", None

        # Check if already completed today
        today = date.today()
        if await self._is_completed_today(owner_type, owner_id, habit_id, today):
            streak = await self.get_habit_streak(habit_id, owner_type, owner_id)
            return False, "Already completed today", {
                "habit": habit,
                "streak": streak,
                "already_completed": True,
            }

        # Create activity record
        # Note: activities table uses 'timestamp', 'kind', and 'habit_name' columns
        from datetime import datetime
        activity_data = {
            "owner_type": owner_type,
            "owner_id": owner_id,
            "habit_id": habit_id,
            "habit_name": habit.get("name", ""),
            "kind": "complete",
            "timestamp": datetime.utcnow().isoformat(),
        }
        
        activity = await self._insert_activity(activity_data)
        
        if not activity:
            return False, "Failed to record completion", None

        # Calculate streak
        streak = await self.get_habit_streak(habit_id, owner_type, owner_id)
        
        return True, "Habit completed", {
            "habit": habit,
            "streak": streak,
            "activity": activity,
        }

    async def get_incomplete_habits_today(
        self,
        owner_id: str,
        owner_type: str = "user",
    ) -> List[Dict[str, Any]]:
        """
        Get list of habits not yet completed today.
        
        Args:
            owner_id: User ID
            owner_type: Type of owner
            
        Returns:
            List of incomplete habit dicts
        """
        habits = await self._get_habits_by_owner(owner_type, owner_id)
        today = date.today()
        
        incomplete = []
        for habit in habits:
            if not habit.get("active", True):
                continue
                
            is_completed = await self._is_completed_today(
                owner_type, owner_id, habit["id"], today
            )
            
            if not is_completed:
                streak = await self.get_habit_streak(habit["id"], owner_type, owner_id)
                incomplete.append({
                    **habit,
                    "completed": False,
                    "streak": streak,
                })
        
        return incomplete

    async def get_all_habits_with_status(
        self,
        owner_id: str,
        owner_type: str = "user",
    ) -> List[Dict[str, Any]]:
        """
        Get all habits with their completion status for today.
        
        Args:
            owner_id: User ID
            owner_type: Type of owner
            
        Returns:
            List of habit dicts with completion status
        """
        habits = await self._get_habits_by_owner(owner_type, owner_id)
        today = date.today()
        
        result = []
        for habit in habits:
            if not habit.get("active", True):
                continue
                
            is_completed = await self._is_completed_today(
                owner_type, owner_id, habit["id"], today
            )
            streak = await self.get_habit_streak(habit["id"], owner_type, owner_id)
            
            # Get goal name if available
            goal_name = None
            if habit.get("goal_id"):
                goal = await self._get_goal_by_id(habit["goal_id"])
                goal_name = goal.get("name") if goal else None
            
            result.append({
                **habit,
                "completed": is_completed,
                "streak": streak,
                "goal_name": goal_name or "No Goal",
            })
        
        return result

    @with_retry()
    async def get_habit_streak(
        self,
        habit_id: str,
        owner_type: str = "user",
        owner_id: Optional[str] = None,
    ) -> int:
        """
        Calculate current streak count for a habit.
        
        Applies retry logic for transient connection errors.
        Requirements: 2.1, 2.3
        
        Note: activities table uses 'timestamp' and 'kind' columns
        instead of 'date' and 'completed'.
        
        Args:
            habit_id: ID of the habit
            owner_type: Type of owner
            owner_id: User ID (optional, for filtering)
            
        Returns:
            Current streak count
        """
        # Get recent activities for this habit, ordered by timestamp descending
        query = self.supabase.table("activities").select("timestamp").eq(
            "habit_id", habit_id
        ).eq("kind", "complete").order("timestamp", desc=True).limit(365)
        
        if owner_id:
            query = query.eq("owner_type", owner_type).eq("owner_id", owner_id)
        
        result = query.execute()
        
        if not result.data:
            return 0

        # Calculate streak by extracting dates from timestamps
        streak = 0
        expected_date = date.today()
        seen_dates = set()
        
        for activity in result.data:
            # Extract date from timestamp (format: 2026-01-11T16:59:28.61+00:00)
            timestamp_str = activity["timestamp"]
            activity_date = date.fromisoformat(timestamp_str[:10])
            
            # Skip if we've already counted this date
            if activity_date in seen_dates:
                continue
            seen_dates.add(activity_date)
            
            # Allow for today or yesterday as the start
            if streak == 0 and activity_date == expected_date:
                streak = 1
                expected_date = expected_date - timedelta(days=1)
            elif streak == 0 and activity_date == expected_date - timedelta(days=1):
                streak = 1
                expected_date = activity_date - timedelta(days=1)
            elif activity_date == expected_date:
                streak += 1
                expected_date = expected_date - timedelta(days=1)
            else:
                break
        
        return streak

    async def get_today_summary(
        self,
        owner_id: str,
        owner_type: str = "user",
    ) -> Dict[str, Any]:
        """
        Get summary of today's habit completion.
        
        Args:
            owner_id: User ID
            owner_type: Type of owner
            
        Returns:
            Summary dict with completed, total, and habits list
        """
        habits = await self.get_all_habits_with_status(owner_id, owner_type)
        
        completed = sum(1 for h in habits if h.get("completed"))
        total = len(habits)
        
        return {
            "completed": completed,
            "total": total,
            "completion_rate": (completed / total * 100) if total > 0 else 0,
            "habits": habits,
        }

    # ========================================================================
    # Private Helper Methods
    # ========================================================================

    @with_retry()
    async def _get_habits_by_owner(
        self,
        owner_type: str,
        owner_id: str,
    ) -> List[Dict[str, Any]]:
        """
        Get all habits for an owner.
        
        Applies retry logic for transient connection errors.
        Requirements: 2.1, 2.3
        """
        result = self.supabase.table("habits").select("*").eq(
            "owner_type", owner_type
        ).eq("owner_id", owner_id).execute()
        
        return result.data if result.data else []

    @with_retry()
    async def _get_habit_by_id(self, habit_id: str) -> Optional[Dict[str, Any]]:
        """
        Get a habit by ID.
        
        Applies retry logic for transient connection errors.
        Requirements: 2.1, 2.3
        """
        result = self.supabase.table("habits").select("*").eq(
            "id", habit_id
        ).execute()
        
        return result.data[0] if result.data else None

    @with_retry()
    async def _get_goal_by_id(self, goal_id: str) -> Optional[Dict[str, Any]]:
        """
        Get a goal by ID.
        
        Applies retry logic for transient connection errors.
        Requirements: 2.1, 2.3
        """
        result = self.supabase.table("goals").select("*").eq(
            "id", goal_id
        ).execute()
        
        return result.data[0] if result.data else None

    @with_retry()
    async def _is_completed_today(
        self,
        owner_type: str,
        owner_id: str,
        habit_id: str,
        check_date: date,
    ) -> bool:
        """
        Check if a habit is completed for a specific date.
        
        Applies retry logic for transient connection errors.
        Requirements: 2.1, 2.3
        
        Note: activities table uses 'timestamp' column and 'kind' column
        instead of 'date' and 'completed'.
        """
        # Filter by date range using timestamp column
        start_of_day = f"{check_date.isoformat()}T00:00:00"
        end_of_day = f"{check_date.isoformat()}T23:59:59"
        
        result = self.supabase.table("activities").select("id").eq(
            "owner_type", owner_type
        ).eq("owner_id", owner_id).eq("habit_id", habit_id).eq(
            "kind", "complete"
        ).gte("timestamp", start_of_day).lte("timestamp", end_of_day).execute()
        
        return len(result.data) > 0 if result.data else False

    @with_retry()
    async def _insert_activity(
        self,
        activity_data: Dict[str, Any],
    ) -> Optional[Dict[str, Any]]:
        """
        Insert an activity record into the database.
        
        Applies retry logic for transient connection errors.
        Requirements: 2.1, 2.3
        
        Args:
            activity_data: Activity data to insert
            
        Returns:
            Inserted activity record or None if failed
        """
        result = self.supabase.table("activities").insert(activity_data).execute()
        return result.data[0] if result.data else None

    def _find_similar_habits(
        self,
        name: str,
        habits: List[Dict[str, Any]],
        threshold: float = 0.6,
    ) -> List[str]:
        """Find habits with similar names."""
        similar = []
        name_lower = name.lower()
        
        for habit in habits:
            habit_name = habit["name"]
            ratio = SequenceMatcher(None, name_lower, habit_name.lower()).ratio()
            if ratio >= threshold:
                similar.append(habit_name)
        
        return similar[:3]  # Return top 3 suggestions

    async def increment_habit_progress(
        self,
        owner_id: str,
        habit_id: str,
        amount: Optional[float] = None,
        source: str = "slack",
        owner_type: str = "user",
    ) -> Tuple[bool, str, Optional[Dict]]:
        """
        Increment habit progress by the specified amount.
        
        Creates an activity record with the specified amount (or the habit's
        workloadPerCount if not specified) and source.
        
        Requirements:
        - 4.2: Create activity record with amount equal to habit's workloadPerCount (default: 1)
        - 4.3: Set source field to "slack" for increment button actions
        
        Args:
            owner_id: User ID
            habit_id: ID of the habit
            amount: Amount to add (defaults to workloadPerCount)
            source: Source of the increment (default: "slack")
            owner_type: Type of owner (default: "user")
            
        Returns:
            Tuple of (success, message, result_data)
            - result_data contains: habit, streak, activity, amount
        """
        # Get habit by ID
        habit = await self._get_habit_by_id(habit_id)
        if not habit:
            return False, "Habit not found", None

        # Get workload_per_count from habit (default to 1 if not set)
        # Note: Supabase column is snake_case: workload_per_count
        workload_per_count = habit.get("workload_per_count", 1)
        if workload_per_count is None:
            workload_per_count = 1

        # If amount parameter is None, use workloadPerCount
        increment_amount = amount if amount is not None else workload_per_count

        # Create activity record with amount and source
        activity_data = {
            "owner_type": owner_type,
            "owner_id": owner_id,
            "habit_id": habit_id,
            "habit_name": habit.get("name", ""),
            "kind": "complete",
            "timestamp": datetime.utcnow().isoformat(),
            "amount": increment_amount,
            "source": source,
        }

        activity = await self._insert_activity(activity_data)

        if not activity:
            return False, "Failed to record progress", None

        # Calculate new streak
        streak = await self.get_habit_streak(habit_id, owner_type, owner_id)

        return True, "Progress updated", {
            "habit": habit,
            "streak": streak,
            "activity": activity,
            "amount": increment_amount,
        }

File: backend/src/services/habitCompletionReporter.ts
/**
 * Habit Completion Reporter
 *
 * Service for processing habit completions from Slack interactions.
 *
 * This service handles habit completion logic and streak calculation,
 * delegating all database operations to injected repositories.
 *
 * Requirements:
 * - 8.1: Complete habit by ID
 * - 8.2: Calculate current streak for a habit
 * - 8.3: Detect duplicate completions
 * - 8.4: Get JST day boundaries
 * - 8.5: Get incomplete habits today
 * - 8.6: Get all habits with status
 *
 * Properties:
 * - Property 9: Streak Calculation - For any habit with activities, the streak should equal
 *   the count of consecutive days (ending today or yesterday) with at least one completion activity.
 * - Property 10: Duplicate Completion Detection - For any habit that has been completed today,
 *   attempting to complete it again should return already_completed=true without creating a new activity.
 */

import type { HabitRepository } from '../repositories/habitRepository.js';
import type { ActivityRepository } from '../repositories/activityRepository.js';
import type { GoalRepository } from '../repositories/goalRepository.js';
import type { Habit, Activity } from '../schemas/habit.js';
import { DataFetchError } from '../errors/index.js';
import { withRetry } from '../utils/retry.js';
import { getLogger } from '../utils/logger.js';

const logger = getLogger('habitCompletionReporter');

/**
 * Result of a habit completion operation.
 */
export interface HabitCompletionResult {
  habit: Habit;
  streak: number;
  activity?: Activity;
  already_completed?: boolean;
}

/**
 * Result of a habit completion attempt.
 */
export type CompletionResult = [boolean, string, HabitCompletionResult | null];

/**
 * Habit with completion status and streak.
 */
export interface HabitWithStatus extends Habit {
  completed: boolean;
  streak: number;
  goal_name?: string | null;
}

/**
 * Habit with suggestions for similar habits.
 */
export interface HabitSuggestions {
  suggestions: string[];
}

/**
 * Today's habit summary.
 */
export interface TodaySummary {
  completed: number;
  total: number;
  completion_rate: number;
  habits: HabitWithStatus[];
}

/**
 * Result of an increment operation.
 */
export interface IncrementResult {
  habit: Habit;
  streak: number;
  activity: Activity;
  amount: number;
}

/**
 * Service for handling habit completions via Slack.
 *
 * This service is responsible for:
 * - Completing habits by name or ID
 * - Calculating habit streaks
 * - Tracking daily completion status
 * - Generating habit summaries
 *
 * All database operations are delegated to injected repositories,
 * following the dependency injection pattern for testability.
 */
export class HabitCompletionReporter {
  private readonly habitRepo: HabitRepository;
  private readonly activityRepo: ActivityRepository;
  private readonly goalRepo: GoalRepository;

  /**
   * Initialize the HabitCompletionReporter with injected repositories.
   *
   * @param habitRepo - Repository for habit database operations.
   * @param activityRepo - Repository for activity database operations.
   * @param goalRepo - Repository for goal database operations.
   */
  constructor(
    habitRepo: HabitRepository,
    activityRepo: ActivityRepository,
    goalRepo: GoalRepository
  ) {
    this.habitRepo = habitRepo;
    this.activityRepo = activityRepo;
    this.goalRepo = goalRepo;
  }

  /**
   * Find and complete a habit by name.
   *
   * Searches for a habit by name (case-insensitive) and completes it if found.
   * If no exact match is found, attempts a partial match. If still not found,
   * returns suggestions for similar habit names.
   *
   * @param ownerId - User ID.
   * @param habitName - Name of the habit to complete.
   * @param source - Source of completion (default: "slack").
   * @param ownerType - Type of owner (default: "user").
   * @returns Tuple of [success, message, habitData].
   */
  async completeHabitByName(
    ownerId: string,
    habitName: string,
    source = 'slack',
    ownerType = 'user'
  ): Promise<[boolean, string, HabitCompletionResult | HabitSuggestions | null]> {
    try {
      // Find habit by name (case-insensitive)
      const habits = await this.habitRepo.getByOwner(ownerType, ownerId);

      // Try exact match first
      let habit = habits.find(
        (h) => h.name.toLowerCase() === habitName.toLowerCase()
      );

      // Try partial match if no exact match
      if (!habit) {
        habit = habits.find((h) =>
          h.name.toLowerCase().includes(habitName.toLowerCase())
        );
      }

      if (!habit) {
        // Find similar habits for suggestions
        const similar = this.findSimilarHabits(habitName, habits);
        if (similar.length > 0) {
          return [false, `Habit '${habitName}' not found`, { suggestions: similar }];
        }
        return [false, `Habit '${habitName}' not found`, null];
      }

      return await this.completeHabitById(ownerId, habit.id, source, ownerType);
    } catch (error) {
      if (error instanceof DataFetchError) {
        throw error;
      }
      throw new DataFetchError(
        `Failed to complete habit by name: ${error instanceof Error ? error.message : String(error)}`,
        error instanceof Error ? error : undefined
      );
    }
  }

  /**
   * Complete a habit by ID.
   *
   * Creates a completion activity for the specified habit if it hasn't
   * already been completed today. Returns the habit details and current streak.
   *
   * **Property 10: Duplicate Completion Detection**
   * For any habit that has been completed today, attempting to complete it again
   * should return already_completed=true without creating a new activity.
   *
   * @param ownerId - User ID.
   * @param habitId - ID of the habit to complete.
   * @param source - Source of completion (default: "slack").
   * @param ownerType - Type of owner (default: "user").
   * @returns Tuple of [success, message, habitData].
   */
  async completeHabitById(
    ownerId: string,
    habitId: string,
    _source = 'slack',
    ownerType = 'user'
  ): Promise<CompletionResult> {
    try {
      // Get habit details using repository
      const habit = await this.habitRepo.getById(habitId);
      if (!habit) {
        return [false, 'Habit not found', null];
      }

      // Check if already completed today using JST boundaries
      const [start, end] = this.getJstDayBoundaries();
      if (await this.activityRepo.hasCompletionToday(habitId, start, end)) {
        const streak = await this.getHabitStreak(habitId, ownerType, ownerId);
        return [
          false,
          'Already completed today',
          {
            habit,
            streak,
            already_completed: true,
          },
        ];
      }

      // Create activity record using repository
      const activityData = {
        owner_type: ownerType,
        owner_id: ownerId,
        habit_id: habitId,
        habit_name: habit.name ?? '',
        kind: 'complete' as const,
        timestamp: new Date().toISOString(),
        amount: habit.workload_per_count ?? 1,
      };

      const activity = await this.activityRepo.create(activityData);

      // Calculate streak
      const streak = await this.getHabitStreak(habitId, ownerType, ownerId);

      logger.info('Habit completed', {
        habit_id: habitId,
        habit_name: habit.name,
        owner_id: ownerId,
        streak,
      });

      return [
        true,
        'Habit completed',
        {
          habit,
          streak,
          activity,
        },
      ];
    } catch (error) {
      if (error instanceof DataFetchError) {
        throw error;
      }
      throw new DataFetchError(
        `Failed to complete habit: ${error instanceof Error ? error.message : String(error)}`,
        error instanceof Error ? error : undefined
      );
    }
  }

  /**
   * Get list of habits not yet completed today.
   *
   * Returns all active habits for the owner that have not been completed
   * today, along with their current streak counts.
   *
   * @param ownerId - User ID.
   * @param ownerType - Type of owner.
   * @returns List of incomplete habit objects with completion status and streak.
   */
  async getIncompleteHabitsToday(
    ownerId: string,
    ownerType = 'user'
  ): Promise<HabitWithStatus[]> {
    try {
      const habits = await this.habitRepo.getByOwner(ownerType, ownerId);
      const [start, end] = this.getJstDayBoundaries();

      const incomplete: HabitWithStatus[] = [];
      for (const habit of habits) {
        if (!habit.active) {
          continue;
        }

        const isCompleted = await this.activityRepo.hasCompletionToday(
          habit.id,
          start,
          end
        );

        if (!isCompleted) {
          const streak = await this.getHabitStreak(habit.id, ownerType, ownerId);
          incomplete.push({
            ...habit,
            completed: false,
            streak,
          });
        }
      }

      return incomplete;
    } catch (error) {
      throw new DataFetchError(
        `Failed to get incomplete habits: ${error instanceof Error ? error.message : String(error)}`,
        error instanceof Error ? error : undefined
      );
    }
  }

  /**
   * Get all habits with their completion status for today.
   *
   * Returns all active habits for the owner with their completion status,
   * streak count, and associated goal name.
   *
   * @param ownerId - User ID.
   * @param ownerType - Type of owner.
   * @returns List of habit objects with completion status, streak, and goal name.
   */
  async getAllHabitsWithStatus(
    ownerId: string,
    ownerType = 'user'
  ): Promise<HabitWithStatus[]> {
    try {
      const habits = await this.habitRepo.getByOwner(ownerType, ownerId);
      const [start, end] = this.getJstDayBoundaries();

      const result: HabitWithStatus[] = [];
      for (const habit of habits) {
        if (!habit.active) {
          continue;
        }

        const isCompleted = await this.activityRepo.hasCompletionToday(
          habit.id,
          start,
          end
        );
        const streak = await this.getHabitStreak(habit.id, ownerType, ownerId);

        // Get goal name if available using repository
        let goalName: string | null = null;
        if (habit.goal_id) {
          const goal = await this.goalRepo.getById(habit.goal_id);
          goalName = goal?.name ?? null;
        }

        result.push({
          ...habit,
          completed: isCompleted,
          streak,
          goal_name: goalName ?? 'No Goal',
        });
      }

      return result;
    } catch (error) {
      throw new DataFetchError(
        `Failed to get habits with status: ${error instanceof Error ? error.message : String(error)}`,
        error instanceof Error ? error : undefined
      );
    }
  }

  /**
   * Calculate current streak count for a habit.
   *
   * Calculates the number of consecutive days the habit has been completed,
   * ending today or yesterday. Uses the activity repository to fetch
   * completion history.
   *
   * **Property 9: Streak Calculation**
   * For any habit with activities, the streak should equal the count of consecutive
   * days (ending today or yesterday) with at least one completion activity.
   *
   * Applies retry logic for transient connection errors.
   *
   * @param habitId - ID of the habit.
   * @param ownerType - Type of owner.
   * @param ownerId - User ID (optional, for filtering).
   * @returns Current streak count (0 if no completions).
   */
  async getHabitStreak(
    habitId: string,
    _ownerType = 'user',
    _ownerId?: string
  ): Promise<number> {
    return withRetry(async () => {
      try {
        // Get recent activities for this habit using repository
        const activities = await this.activityRepo.getHabitActivities(
          habitId,
          'complete',
          365
        );

        if (activities.length === 0) {
          return 0;
        }

        // Calculate streak by extracting dates from timestamps
        let streak = 0;
        let expectedDate = this.getToday();
        const seenDates = new Set<string>();

        for (const activity of activities) {
          // Extract date from timestamp (format: 2026-01-11T16:59:28.61+00:00)
          const timestampStr = activity.timestamp;
          const activityDateStr = timestampStr.slice(0, 10);
          const activityDate = new Date(activityDateStr);

          // Skip if we've already counted this date
          if (seenDates.has(activityDateStr)) {
            continue;
          }
          seenDates.add(activityDateStr);

          // Allow for today or yesterday as the start
          if (streak === 0 && this.isSameDate(activityDate, expectedDate)) {
            streak = 1;
            expectedDate = this.subtractDays(expectedDate, 1);
          } else if (
            streak === 0 &&
            this.isSameDate(activityDate, this.subtractDays(expectedDate, 1))
          ) {
            streak = 1;
            expectedDate = this.subtractDays(activityDate, 1);
          } else if (this.isSameDate(activityDate, expectedDate)) {
            streak += 1;
            expectedDate = this.subtractDays(expectedDate, 1);
          } else {
            break;
          }
        }

        return streak;
      } catch (error) {
        throw new DataFetchError(
          `Failed to calculate streak: ${error instanceof Error ? error.message : String(error)}`,
          error instanceof Error ? error : undefined
        );
      }
    });
  }

  /**
   * Get summary of today's habit completion.
   *
   * Returns a summary including the number of completed habits,
   * total habits, completion rate, and detailed habit list.
   *
   * @param ownerId - User ID.
   * @param ownerType - Type of owner.
   * @returns Summary object with completed, total, completion_rate, and habits list.
   */
  async getTodaySummary(ownerId: string, ownerType = 'user'): Promise<TodaySummary> {
    try {
      const habits = await this.getAllHabitsWithStatus(ownerId, ownerType);

      const completed = habits.filter((h) => h.completed).length;
      const total = habits.length;

      return {
        completed,
        total,
        completion_rate: total > 0 ? (completed / total) * 100 : 0,
        habits,
      };
    } catch (error) {
      throw new DataFetchError(
        `Failed to get today summary: ${error instanceof Error ? error.message : String(error)}`,
        error instanceof Error ? error : undefined
      );
    }
  }

  /**
   * Increment habit progress by the specified amount.
   *
   * Creates an activity record with the specified amount (or the habit's
   * workloadPerCount if not specified) and source.
   *
   * @param ownerId - User ID.
   * @param habitId - ID of the habit.
   * @param amount - Amount to add (defaults to workloadPerCount).
   * @param source - Source of the increment (default: "slack").
   * @param ownerType - Type of owner (default: "user").
   * @returns Tuple of [success, message, resultData].
   */
  async incrementHabitProgress(
    ownerId: string,
    habitId: string,
    amount?: number,
    _source = 'slack',
    ownerType = 'user'
  ): Promise<[boolean, string, IncrementResult | null]> {
    try {
      // Get habit by ID using repository
      const habit = await this.habitRepo.getById(habitId);
      if (!habit) {
        return [false, 'Habit not found', null];
      }

      // Get workload_per_count from habit (default to 1 if not set)
      const workloadPerCount = habit.workload_per_count ?? 1;

      // If amount parameter is undefined, use workloadPerCount
      const incrementAmount = amount ?? workloadPerCount;

      // Create activity record with amount using repository
      const activityData = {
        owner_type: ownerType,
        owner_id: ownerId,
        habit_id: habitId,
        habit_name: habit.name ?? '',
        kind: 'complete' as const,
        timestamp: new Date().toISOString(),
        amount: incrementAmount,
      };

      const activity = await this.activityRepo.create(activityData);

      // Calculate new streak
      const streak = await this.getHabitStreak(habitId, ownerType, ownerId);

      logger.info('Habit progress incremented', {
        habit_id: habitId,
        habit_name: habit.name,
        owner_id: ownerId,
        amount: incrementAmount,
        streak,
      });

      return [
        true,
        'Progress updated',
        {
          habit,
          streak,
          activity,
          amount: incrementAmount,
        },
      ];
    } catch (error) {
      throw new DataFetchError(
        `Failed to increment habit progress: ${error instanceof Error ? error.message : String(error)}`,
        error instanceof Error ? error : undefined
      );
    }
  }

  // ========================================================================
  // Private Helper Methods
  // ========================================================================

  /**
   * Get JST day start and end times.
   *
   * Returns the start (00:00:00) and end (23:59:59) times for the current
   * day in Japan Standard Time.
   *
   * @returns Tuple of [startDatetime, endDatetime] in JST.
   */
  getJstDayBoundaries(): [Date, Date] {
    // Get current time in JST
    const now = new Date();
    const jstOffset = 9 * 60; // JST is UTC+9
    const utcOffset = now.getTimezoneOffset();
    const jstTime = new Date(now.getTime() + (jstOffset + utcOffset) * 60 * 1000);

    // Create start of day in JST
    const startJst = new Date(jstTime);
    startJst.setHours(0, 0, 0, 0);

    // Create end of day in JST
    const endJst = new Date(jstTime);
    endJst.setHours(23, 59, 59, 999);

    // Convert back to UTC for database queries
    const start = new Date(startJst.getTime() - (jstOffset + utcOffset) * 60 * 1000);
    const end = new Date(endJst.getTime() - (jstOffset + utcOffset) * 60 * 1000);

    return [start, end];
  }

  /**
   * Get today's date (date only, no time).
   *
   * @returns Today's date at midnight.
   */
  private getToday(): Date {
    const today = new Date();
    today.setHours(0, 0, 0, 0);
    return today;
  }

  /**
   * Subtract days from a date.
   *
   * @param date - The date to subtract from.
   * @param days - Number of days to subtract.
   * @returns New date with days subtracted.
   */
  private subtractDays(date: Date, days: number): Date {
    const result = new Date(date);
    result.setDate(result.getDate() - days);
    return result;
  }

  /**
   * Check if two dates are the same (ignoring time).
   *
   * @param date1 - First date.
   * @param date2 - Second date.
   * @returns True if dates are the same day.
   */
  private isSameDate(date1: Date, date2: Date): boolean {
    return (
      date1.getFullYear() === date2.getFullYear() &&
      date1.getMonth() === date2.getMonth() &&
      date1.getDate() === date2.getDate()
    );
  }

  /**
   * Find habits with similar names.
   *
   * Uses sequence matching to find habits with names similar to the
   * provided name, useful for providing suggestions when a habit is not found.
   *
   * @param name - The name to search for.
   * @param habits - List of habit objects to search.
   * @param threshold - Minimum similarity ratio (0.0 to 1.0). Defaults to 0.6.
   * @returns List of similar habit names (up to 3 suggestions).
   */
  private findSimilarHabits(
    name: string,
    habits: Habit[],
    threshold = 0.6
  ): string[] {
    const similar: string[] = [];
    const nameLower = name.toLowerCase();

    for (const habit of habits) {
      const habitName = habit.name;
      const ratio = this.calculateSimilarity(nameLower, habitName.toLowerCase());
      if (ratio >= threshold) {
        similar.push(habitName);
      }
    }

    return similar.slice(0, 3); // Return top 3 suggestions
  }

  /**
   * Calculate similarity ratio between two strings.
   *
   * Uses a simple Levenshtein-based similarity calculation.
   *
   * @param str1 - First string.
   * @param str2 - Second string.
   * @returns Similarity ratio between 0 and 1.
   */
  private calculateSimilarity(str1: string, str2: string): number {
    const longer = str1.length > str2.length ? str1 : str2;
    const shorter = str1.length > str2.length ? str2 : str1;

    if (longer.length === 0) {
      return 1.0;
    }

    const editDistance = this.levenshteinDistance(longer, shorter);
    return (longer.length - editDistance) / longer.length;
  }

  /**
   * Calculate Levenshtein distance between two strings.
   *
   * @param str1 - First string.
   * @param str2 - Second string.
   * @returns Edit distance between the strings.
   */
  private levenshteinDistance(str1: string, str2: string): number {
    const m = str1.length;
    const n = str2.length;

    // Create a 2D array with proper initialization
    const matrix: number[][] = Array.from({ length: n + 1 }, () =>
      Array.from({ length: m + 1 }, () => 0)
    );

    // Initialize first column
    for (let i = 0; i <= n; i++) {
      matrix[i]![0] = i;
    }

    // Initialize first row
    for (let j = 0; j <= m; j++) {
      matrix[0]![j] = j;
    }

    // Fill in the rest of the matrix
    for (let i = 1; i <= n; i++) {
      for (let j = 1; j <= m; j++) {
        if (str2.charAt(i - 1) === str1.charAt(j - 1)) {
          matrix[i]![j] = matrix[i - 1]![j - 1]!;
        } else {
          matrix[i]![j] = Math.min(
            matrix[i - 1]![j - 1]! + 1, // substitution
            matrix[i]![j - 1]! + 1, // insertion
            matrix[i - 1]![j]! + 1 // deletion
          );
        }
      }
    }

    return matrix[n]![m]!;
  }
}
